unstructured_ingest/v2/processes/connectors/github.py

```diff
@@ -0,0 +1,343 @@
+        return values
+
+    @SourceConnectionError.wrap
+    @requires_dependencies(["github"], extras="github")
+    def get_repo(self) -> "Repository":
+        from github import Github
+
+        github = Github(self.access_config.get_secret_value().git_access_token)
+        return github.get_repo(self.repo_path)
+
+
+class GitHubIndexerConfig(IndexerConfig):
+    pass
+
+
+@dataclass
+class GitHubIndexer(Indexer):
+    connection_config: GitHubConnectionConfig
+    index_config: GitHubIndexerConfig
+
+    def precheck(self) -> None:
+        """Performs a precheck to validate the connection to the GitHub repository.
+
+        This method sends a `HEAD` request to the GitHub API to ensure the repository
+        is accessible and properly configured. It uses the GitHub `Requester` class
+        with retry support and authentication via an access token.
+
+        Raises:
+            SourceConnectionError: If the connection validation fails.
+        """
+        from github import Auth, Consts
+        from github.GithubRetry import GithubRetry
+        from github.Requester import Requester
+
+        auth = Auth.Token(self.connection_config.access_config.get_secret_value().git_access_token)
+
+        try:
+            requester = Requester(
+                auth=auth,
+                base_url=Consts.DEFAULT_BASE_URL,
+                timeout=Consts.DEFAULT_TIMEOUT,
+                user_agent=Consts.DEFAULT_USER_AGENT,
+                per_page=Consts.DEFAULT_PER_PAGE,
+                verify=True,
+                retry=GithubRetry(),
+                pool_size=None,
+            )
+            url = f"{Consts.DEFAULT_BASE_URL}/repos/{self.connection_config.repo_path}"
+            logger.debug(f"Precheck Request: {url!r}")
+
+            headers, _ = requester.requestJsonAndCheck("HEAD", url)
+            logger.debug(f"Headers from HEAD request: {headers}")
+        except Exception as e:
+            logger.error(f"Failed to validate connection: {e}", exc_info=True)
+            raise SourceConnectionError(f"Failed to validate connection: {e}")
+
+    def run(self, **kwargs: Any) -> Generator[FileData, None, None]:
+        """Iterates over the GitHub repository tree, yielding `FileData` objects for all
+        files (blobs).
+
+        This method retrieves the entire repository tree for the specified branch or
+        the default branch.
+        For each file (blob), it extracts relevant metadata and yields a `FileData` object.
+
+        Args:
+            **kwargs (Any): Additional optional arguments.
+
+        Yields:
+            FileData: An object containing metadata and identifiers for each file in the repository.
+        """
+        repo = self.connection_config.get_repo()
+        sha = self.connection_config.branch or repo.default_branch
+        logger.info(f"Starting to look for blob files on GitHub in branch: {sha!r}")
+
+        git_tree = repo.get_git_tree(sha, recursive=True)
```

Comment:
Does it make sense to have the recursive flag part of the indexer config? This is the usual pattern we've taken for connectors that support recursive files. Not sure what this would pull in without the recursive call. Is there a way to specificy a file path within a github repo to pull files in?

Reply:
Hey Roman!

I’ve explored and tested some alternatives. In the library, we have two relevant methods:

- [get_contents](https://pygithub.readthedocs.io/en/stable/github_objects/Repository.html?highlight=get_contents#github.Repository.Repository.get_contents)
- [get_git_tree](https://pygithub.readthedocs.io/en/stable/github_objects/Repository.html?highlight=get_contents#github.Repository.Repository.get_git_tree)

With `get_contents`, you can specify a particular directory, but it won’t read the subfolders within that path. To maintain the same behavior as GitHub API v1 (which supports recursion), I think we should stick with this approach.

What do you think?

Reply:
Yep, lets maintain the given behavior.

Reply:
I may have misstated my response, can we still map this `recursive` flag to an indexer config?

Reply:
As discussed in the private channel, I’ve added a new recursive flag to the Indexer.

Reply:
As discussed in the private channel, I’ve added a new recursive flag to the Indexer.